packages/intrepid2/unit-test/performance/StructuredIntegration/StructuredIntegrationPerformance.cpp

```diff
@@ -0,0 +1,953 @@
+  return cellStiffness;
+}
+
+//! returns an estimated count of the floating point operations performed.
+template<class Scalar, class PointScalar, int spaceDim, typename ExecSpaceType>
+void performStructuredQuadratureHypercubeGRADGRAD(CellGeometry<PointScalar, spaceDim, ExecSpaceType> &geometry, const int &polyOrder, const int &worksetSize,
+                                                  double &transformIntegrateFlopCount, double &jacobianCellMeasureFlopCount)
+{
+  int numVertices = 1;
+  for (int d=0; d<spaceDim; d++)
+  {
+    numVertices *= 2;
+  }
+  
+  auto initialSetupTimer = Teuchos::TimeMonitor::getNewTimer("Initial Setup");
+  initialSetupTimer->start();
+  using namespace std;
+  using FunctionSpaceTools = Intrepid2::FunctionSpaceTools<ExecSpaceType>;
+  using IntegrationTools   = Intrepid2::IntegrationTools<ExecSpaceType>;
+  // dimensions of the returned view are (C,F,F)
+  auto fs = Intrepid2::FUNCTION_SPACE_HGRAD;
+  
+  shards::CellTopology cellTopo = geometry.cellTopology();
+  
+  auto basis = Intrepid2::getBasis< Intrepid2::DerivedNodalBasisFamily<> >(cellTopo, fs, polyOrder);
+  
+  int numFields = basis->getCardinality();
+  int numCells = geometry.numCells();
+    
+  // local stiffness matrix:
+  ScalarView<Scalar,ExecSpaceType> cellStiffness("cell stiffness matrices",numCells,numFields,numFields);
+  
+  auto cubature = Intrepid2::DefaultCubatureFactory::create<ExecSpaceType>(cellTopo,polyOrder*2);
+  auto tensorCubatureWeights = cubature->allocateTensorCubatureWeights();
+  TensorPoints<PointScalar,ExecSpaceType> tensorCubaturePoints  = cubature->allocateTensorCubaturePoints();
+  
+  cubature->getTensorCubature(tensorCubaturePoints, tensorCubatureWeights);
+  
+  EOperator op = OPERATOR_GRAD;
+  BasisValues<Scalar,ExecSpaceType> gradientValues = basis->allocateBasisValues(tensorCubaturePoints, op);
+  basis->getValues(gradientValues, tensorCubaturePoints, op);
+  
+  // goal here is to do a weighted Poisson; i.e. (f grad u, grad v) on each cell
+    
+  int cellOffset = 0;
+  
+  auto jacobianAndCellMeasureTimer = Teuchos::TimeMonitor::getNewTimer("Jacobians");
+  auto fstIntegrateCall = Teuchos::TimeMonitor::getNewTimer("transform + integrate()");
+  
+  Data<PointScalar,ExecSpaceType> jacobian = geometry.allocateJacobianData(tensorCubaturePoints, 0, worksetSize);
+  Data<PointScalar,ExecSpaceType> jacobianDet = CellTools<ExecSpaceType>::allocateJacobianDet(jacobian);
+  Data<PointScalar,ExecSpaceType> jacobianInv = CellTools<ExecSpaceType>::allocateJacobianInv(jacobian);
+  TensorData<PointScalar,ExecSpaceType> cellMeasures = geometry.allocateCellMeasure(jacobianDet, tensorCubatureWeights);
+  
+  // lazily-evaluated transformed gradient values (temporary to allow integralData allocation)
+  auto transformedGradientValuesTemp = FunctionSpaceTools::getHGRADtransformGRAD(jacobianInv, gradientValues.vectorData());
```

Comment:
now that we have the geometry class, it would be nice if the transformations  (e.g. `getHGradTransformHGrad`) would take as argument the geometry instead of a subset of {Jacobian, JacbianInv, JacobianDet}, so that they would all look the same. And we could also implement the inverse of the pullBack (pushForward?) needed for projections.

Reply:
Yes, I see what you're aiming at, and I agree that would be nice; we could even have a generic pullback call that takes a function space and an operator.  The one issue I see is that on a naive implementation, you'd end up needing to allocate and compute the Jacobian, etc. on each transformation call, whereas the current implementation allows reuse.  We could address that by storing a lazily-computed Jacobian, etc. in the CellGeometry object.  We can talk this through to figure out the right design.